### ISSUE TITLE
Can't apply explain_weights neither show_weights for sklearn_crfsuit.CRF object

### ISSUE THREAD

**Issue by davidsbatista** [TeamHG-Memex/eli5#242]
I was following this tutorial

https://github.com/TeamHG-Memex/eli5/blob/master/docs/source/_notebooks/debug-sklearn-crfsuite.rst

and got an error when trying to get the weights and features of the trained CRF model.

I can't get apply `explain_weights` neither `show_weights` for sklearn_crfsuite.CRF object.

I'm running all inside a python Jupyter notebook:

```eli5.explain_weights_sklearn_crfsuite(crf)```

```---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-75-2efd5261dc5d> in <module>()
      1 import eli5
----> 2 eli5.explain_weights_sklearn_crfsuite(crf)

~/virtual_envs/python3/lib/python3.6/site-packages/eli5/sklearn_crfsuite/explain_weights.py in explain_weights_sklearn_crfsuite(crf, top, target_names, targets, feature_re, feature_filter)
     26     ``feature_re`` and ``feature_filter`` parameters.
     27     """
---> 28     feature_names = np.array(crf.attributes_)
     29     state_coef = crf_state_coef(crf).todense().A
     30     transition_coef = crf_transition_coef(crf)

~/virtual_envs/python3/lib/python3.6/site-packages/sklearn_crfsuite/estimator.py in attributes_(self)
    463         A list of known attributes.
    464         """
--> 465         if self._info is None:
    466             return None
    467 

~/virtual_envs/python3/lib/python3.6/site-packages/sklearn_crfsuite/estimator.py in _info(self)
    497             return None
    498         if self._info_cached is None:
--> 499             self._info_cached = self.tagger_.info()
    500         return self._info_cached
    501 

pycrfsuite/_pycrfsuite.pyx in pycrfsuite._pycrfsuite.Tagger.info (pycrfsuite/_pycrfsuite.cpp:9581)()

pycrfsuite/_pycrfsuite.pyx in pycrfsuite._pycrfsuite.Tagger.info (pycrfsuite/_pycrfsuite.cpp:9498)()

~/virtual_envs/python3/lib/python3.6/site-packages/pycrfsuite/_dumpparser.py in feed(self, line)
     60             self.state = None
     61         else:
---> 62             getattr(self, 'parse_%s' % self.state)(line)
     63 
     64     def parse_FILEHEADER(self, line):

~/virtual_envs/python3/lib/python3.6/site-packages/pycrfsuite/_dumpparser.py in parse_STATE_FEATURES(self, line)
     83     def parse_STATE_FEATURES(self, line):
     84         m = re.match(r"\(\d+\) (.+) --> (.+): ([+-]?\d+\.\d+)", line)
---> 85         attr, label = m.group(1), m.group(2)
     86         assert attr in self.result.attributes
     87         assert label in self.result.labels

AttributeError: 'NoneType' object has no attribute 'group'
```

and with `show_weigths` I have a similar error

`eli5.show_weights(crf, top=30)`

```---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-76-ed63fb41bd58> in <module>()
----> 1 eli5.show_weights(crf, top=30)

~/virtual_envs/python3/lib/python3.6/site-packages/eli5/ipython.py in show_weights(estimator, **kwargs)
    116     """
    117     format_kwargs, explain_kwargs = _split_kwargs(kwargs)
--> 118     expl = explain_weights(estimator, **explain_kwargs)
    119     html = format_as_html(expl, **format_kwargs)
    120     return HTML(html)

~/virtual_envs/python3/lib/python3.6/site-packages/singledispatch.py in wrapper(*args, **kw)
    208 
    209     def wrapper(*args, **kw):
--> 210         return dispatch(args[0].__class__)(*args, **kw)
    211 
    212     registry[object] = func

~/virtual_envs/python3/lib/python3.6/site-packages/eli5/sklearn_crfsuite/explain_weights.py in explain_weights_sklearn_crfsuite(crf, top, target_names, targets, feature_re, feature_filter)
     26     ``feature_re`` and ``feature_filter`` parameters.
     27     """
---> 28     feature_names = np.array(crf.attributes_)
     29     state_coef = crf_state_coef(crf).todense().A
     30     transition_coef = crf_transition_coef(crf)

~/virtual_envs/python3/lib/python3.6/site-packages/sklearn_crfsuite/estimator.py in attributes_(self)
    463         A list of known attributes.
    464         """
--> 465         if self._info is None:
    466             return None
    467 

~/virtual_envs/python3/lib/python3.6/site-packages/sklearn_crfsuite/estimator.py in _info(self)
    497             return None
    498         if self._info_cached is None:
--> 499             self._info_cached = self.tagger_.info()
    500         return self._info_cached
    501 

pycrfsuite/_pycrfsuite.pyx in pycrfsuite._pycrfsuite.Tagger.info (pycrfsuite/_pycrfsuite.cpp:9581)()

pycrfsuite/_pycrfsuite.pyx in pycrfsuite._pycrfsuite.Tagger.info (pycrfsuite/_pycrfsuite.cpp:9498)()

~/virtual_envs/python3/lib/python3.6/site-packages/pycrfsuite/_dumpparser.py in feed(self, line)
     60             self.state = None
     61         else:
---> 62             getattr(self, 'parse_%s' % self.state)(line)
     63 
     64     def parse_FILEHEADER(self, line):

~/virtual_envs/python3/lib/python3.6/site-packages/pycrfsuite/_dumpparser.py in parse_STATE_FEATURES(self, line)
     83     def parse_STATE_FEATURES(self, line):
     84         m = re.match(r"\(\d+\) (.+) --> (.+): ([+-]?\d+\.\d+)", line)
---> 85         attr, label = m.group(1), m.group(2)
     86         assert attr in self.result.attributes
     87         assert label in self.result.labels

AttributeError: 'NoneType' object has no attribute 'group'
```

Nevertheless if I call `crf.tagger_.dump(filename="crf_tagger.txt")` I get a file `crf_tagger.txt` with all the information about features, state transitions, etc.

**Comment by kmike** [comment]
Hm, I just reinstalled everything, and tutorial works for me. I'm not sure how to debug this. Could you try updating python-crfsuite? What package versions are you using (sklearn-crfsuite, python-crfsuite, eli5)?

**Comment by davidsbatista** [comment]
sklearn-crfsuite (0.3.6)
python-crfsuite (0.9.2)
eli5 (0.7)

everything installed through pip

**Comment by kmike** [comment]
Is it failing for the tutorial as-is, or have you changed anything? It shouldn't fail with changes; I'm just trying to figure out how to reproduce it. 

It looks like a bug in https://github.com/scrapinghub/python-crfsuite - it has code which parses `crf.tagger_.dump(filename="crf_tagger.txt")` output to get data in a structured format, and this fails for you.

**Comment by kmike** [comment]
what is your NLTK version, by the way?

**Comment by davidsbatista** [comment]
I'm using the tutorial code, but with a different corpus and roughly the same feature extraction temples just some small changes.

I don't have NLTK installed, is it a dependency?

Also, using: Python 3.6.2 (default, Jul 17 2017, 16:44:45)

**Comment by kmike** [comment]
NLTK is not a dependency for eli5, but it is required to run the tutorial (corpus is from NLTK).

I suspect corpus can trigger the bug; feature names extracted on your data may break the export. Export is just formatted text, and I'm not sure data is escaped properly by crfsuite. It looks very similar to https://github.com/scrapinghub/python-crfsuite/issues/14.

**Comment by davidsbatista** [comment]
I'm now just reducing the feature extraction template for a simple set of features, I will let you know how it goes. Thanks for the quick feedback!

**Comment by davidsbatista** [comment]
I've noticed that in the `crf.tagger_.dump(filename="crf_tagger.txt")` dumped file, I have long line breaks, this is happens due to tokenisation of the input text, producing several `\n` as a token, e.g.:

```
     24: +1:word.lower():,
     25: +1:shape:,
     26: +2:word.lower():




     27: +2:shape:,
     28: word.lower():,
     29: shape:,
     30: -1:word.lower():hallo
     31: -2:word.lower():flüge
     32: -2:word.istitle()
     33: -2:word.isupper()
     34: -2:word.isdigit()
     35: -2:shape:Xxxxx
     36: +1:word.lower():




     37: +1:shape:




     38: +2:word.lower():ich
     39: +2:shape:
```

might this be the reason for the regex to fail?

**Comment by davidsbatista** [comment]
It was indeed that the reason! Fixed that and it's working now! =)

**Comment by kmike** [comment]
Yeah, it looks like a reason. Ideally, it should be fixed in python-crfsuite (https://github.com/scrapinghub/python-crfsuite/issues/14). Thanks for confirming it!

**Comment by tteguayco** [comment]
Where do you guys are exactly performing the replacements of new lines and spaces? It seems I am having the same problem... Thanks in advance :-)

**Comment by wdechanelo** [comment]
> Where are you guys exactly performing the replacements of new lines and spaces? It seems I am having the same problem... Thanks in advance :-)

**Comment by deepakiim** [comment]
> It was indeed that the reason! Fixed that, basically replacing new lines and spaces by #NEWLINE and #SPACE and it's working now! =)

This is still an issue and replacing \n characters or \s is not a solution. Since I need to find the original position of predicted token in original text. When i replace \n with in original text. This is no longer possible.

Where are you guys exactly performing the replacements of new lines and spaces? 
TIA
